internal/repository/db.go

```diff
@@ -0,0 +1,182 @@
+package repository
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/jackc/pgx/v5/pgconn"
+	"github.com/javaman/go-metrics/internal/model"
+)
+
+type databaseStorage struct {
+	db *sql.DB
+	tx *sql.Tx
+}
+
+func retry(f func() error) error {
+	sleep := [...]int{1, 3, 5}
+	var err error
+	for _, s := range sleep {
+		err = f()
+		var netError net.Error
+		if err == nil || !errors.As(err, &netError) || pgconn.SafeToRetry(err) {
+			break
+		}
+		fmt.Println(err)
+		time.Sleep(time.Duration(s) * time.Second)
+	}
+	return err
+}
+
+func createTable(db *sql.DB) error {
+	return retry(func() error {
+		_, err := db.Exec(`
+		CREATE TABLE IF NOT EXISTS metrics (
+			id text,
+			mtype text,
+			delta bigint,
+			value double precision,
+			CONSTRAINT metrics_pk PRIMARY KEY(id, mtype)
+		)
+	        `)
+		return err
+	})
+}
+
+func (d *databaseStorage) get(id string, mtype string) (int64, float64, bool, error) {
+	queryFunc := d.db.QueryRow
+	if d.tx != nil {
+		queryFunc = d.tx.QueryRow
+	}
+
+	var delta int64
+	var value float64
+
+	err := retry(func() error {
+		return queryFunc("SELECT delta, value FROM metrics WHERE id=$1 and mtype=$2", id, mtype).Scan(&delta, &value)
+	})
+
+	switch {
+	case err == sql.ErrNoRows:
+		return delta, value, false, nil
+	case err != nil:
+		return 0, 0, false, err
+	default:
+		return delta, value, true, nil
+	}
+}
+
+func (d *databaseStorage) getAll(mtype string, consumer func(string, int64, float64) bool) error {
```

Comment:
давай договоримся никогда не передавать функции в аргументы. Очень редко когда это оправдано. Такой код очень неочивидный и практически нечитаем
```suggestion
func (d *databaseStorage) getAll(mtype string) ([]Metric, error) {
```